GCEWindowsAgent/serial_listener.go

```diff
@@ -0,0 +1,462 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"syscall"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	"github.com/hashicorp/golang-lru"
+	"github.com/tarm/serial"
+)
+
+type snapshotConfig struct {
+	Timeout               int // seconds
+	ContinueOnScriptError bool
+	PreSnapshotScriptUrl  string
+	PostSnapshotScriptUrl string
+}
+
+type agentReady struct {
+	Identifier string `json:"identifier"`
+	Signature  string `json:"signature"`
+	Version    int    `json:"version"`
+}
+
+type agentShutdown struct {
+	Identifier string `json:"identifier"`
+	Signature  string `json:"signature"`
+	Version    int    `json:"version"`
+}
+
+type serialRequest struct {
+	Identifier  string `json:"identifier"`
+	Signature   string `json:"signature"`
+	Version     int    `json:"version"`
+	OperationId int    `json:"operation_id"`
+	AllDisks    bool   `json:"all_disks"`
+	Disks       string `json:"disks"`
+}
+
+type serialResponse struct {
+	Identifier  string `json:"identifier"`
+	Signature   string `json:"signature"`
+	Version     int    `json:"version"`
+	Rc          int    `json:"rc"`
+	OperationId int    `json:"operation_id"`
+}
+
+const identifier = "329d1643-127b-4096-84a6-ac19f597e51c"
+const testPreSnapshotScript = "preSnapshotScript.sh"
+const testPostSnapshotScript = "postSnapshotScript.sh"
+const serialPort = "/dev/ttyS3"
+const serialProtocolVersion = 1
+
+var (
+	seenPreSnapshotoperationIds, _  = lru.New(128)
+	seenPostSnapshotoperationIds, _ = lru.New(128)
+	currentMsg                      []byte
+	metadataAttributesURL           = "http://metadata.google.internal/computeMetadata/v1/project/attributes/"
+	snapshotMetadataAttributes      = []string{
+		"snapshot-timeout-in-seconds",
+		"snapshot-continue-on-script-error",
+		"snapshot-pre-snapshot-script-url",
+		"snapshot-post-snapshot-script-url"}
+
+	storageURL = "storage.googleapis.com"
+
+	bucket = `([a-z0-9][-_.a-z0-9]*)`
+	object = `(.+)`
+	// Many of the Google Storage URLs are supported below.
+	// It is preferred that customers specify their object using
+	// its gs://<bucket>/<object> URL.
+	bucketRegex = regexp.MustCompile(fmt.Sprintf(`^gs://%s/?$`, bucket))
+	gsRegex     = regexp.MustCompile(fmt.Sprintf(`^gs://%s/%s$`, bucket, object))
+	// Check for the Google Storage URLs:
+	// http://<bucket>.storage.googleapis.com/<object>
+	// https://<bucket>.storage.googleapis.com/<object>
+	gsHTTPRegex1 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://%s\.storage\.googleapis\.com/%s$`, bucket, object))
```

Comment:
why should the agent need to have local hardcoded understanding of all possible gs paths? gcs team owns these definitions and should offer us some parser, or else we should restrict customer to specific path types. i don't want to maintain these (or have any regex in this agent at all, if possible)